#15 farm buildings

keyz182 opened the pull request:
~Not ready yet - initial pass adding a battery farm.~
~I've stripped out the fuel consuption logic for now to simplify while debugging interaction.~
~Current issue is that food can't be hauled to the farm, and I'm a bit stuck on the Work/Jobs system. Tried to take inspiration from the Draincasket from VRE-Sanguophage, but still have hit a wall.~

keyz182 commented:
I believe this is ready for review/QA. 

Needs some input on build cost, and should it be powered or is the food input enough?

Farm will accept up to 16 animals. You attach hoppers to the farm to load in food. Consumption is calculated based on the animals contained, and is triggered daily. If the animals are appropriately fed that day, they produce once per day (if they can produce). At the moment, the output is just placed on the ground for collection.

If animals are unfed, they don't die, they just don't produce. 

I also added a cheekey little easter egg - there's a "Release Animals" button, and an "Eject Animals" button.

feldoh commented on 1.5/Source/PawnStorages/PawnStorages/Farm/CompFarmStorageRefuelable.cs:
```diff
@@ -0,0 +1,137 @@
+        public override string CompInspectStringExtra()
+        {
+            StringBuilder sb = new(base.CompInspectStringExtra());
+            sb.Append("\nTODO: Add nutrition deets");
```
Todo

feldoh commented on 1.5/Defs/Buildings/Buildings_Structure.xml:
```diff
@@ -120,5 +115,116 @@
+						<li>AllowPlantFood</li>
+					</specialFiltersToDisallow>
+					<disallowedThingDefs>
+						<li>Hay</li>
```
why disallow hay and plant food?

feldoh commented on 1.5/Defs/Buildings/Buildings_Structure.xml:
```diff
@@ -120,5 +115,116 @@
+			<Mass>30</Mass>
+		</statBases>
+		<costList>
+			<Steel>15</Steel>
```
that seems way too cheap!

feldoh reviewed (commented):
I think you might not have pushed latest because I can't see anything in here that would spawn the animal products?

Balancing thoughts:
* Cost should include some wood as well as the steel and maybe some hay or something
* This is meant to be significantly slower than leaving them all out there so it's a fair trade. Once per day is the same rate as common stuff like milk and far more common making this way more efficient than actually farming. Imo it should just be a scaled factor on the normal settings defaulting to say once per quadrum or something. There's shearIntervalDays and milkIntervalDays on the compmilkable / shearable or more generally CompHasGatherableBodyResource has a GatherResourcesIntervalDays. I'm happy with only updating once per day but I'd suggest basically updating the fullness on those comps directly, just only once per day so if they're let out it doesn't reset. Alternately trying to build up a dict of `($ResourceDef_$ResourceAmount_$GatherResourcesIntervalDays)` -> animals then the more of an animal which produces x every y days it can be divided down so 2x (Milk => Every 1 days) means 1 milk every half day. Then 1 day divided by half day gets you two so you can produce two batches per day.
* Nutrition cost should also scale with baseHungerRate of the animals included so you cant store a thrumbo on half an ear of corn :p

keyz182 commented:
Looks like I introduced a bug before committing, the job to take an animal to the farm is broken somehow. It's listed twice in the context menu, and doesn't actually do anything. No time tonight to look into it though.

keyz182 commented:
Right, latest push - an approximation of the normal game hunger mechanics. If left hungry, pawns get malnutrition. If it gets to 100% severity, they die. 

Production speed it now based on the normal mechanics too, taking hunger/health into account. 

Should be trivial now to add in some randomized scaling to increase the time between productions. 

Question - Should we produce whenever ready, or keep a list of produced things, and once a day spawn them in bulk? One set of hauling jobs vs a slow stream through the day _might_ be better?

keyz182 commented:
Scaling of the produce interval added. It's now doubled, but configurable in the defs. 

I think all that's left before it's ready is to figure out the buggieness in the order patching that's causing doubling up of jobs.

keyz182 commented:
And there's a bug where pawns that die of starvation don't spawn back properly.

feldoh commented:
I think you're right that once per day is better. I also think the malnutrition/starvation should be optional like it is for the others. 

There's a bool in settings `AllowNeedsDrop` which lets the user make that choice. Extra perf for less real farming. I think no outputs with no food should always apply just wrap the needs bit in `if (PawnStoragesMod.settings.AllowNeedsDrop)`.

feldoh commented:
> And there's a bug where pawns that die of starvation don't spawn back properly.

I don't think we should allow them to die in storage. I saw in the code you have an auto-release, I love that. Could be optional via gizmo if you like but if the animals get to 0 food and can't be fed, just release them to graze seems like a nice option. I don't think they should get to 100% malnutrition before release though because I think that straight up kills them.

keyz182 commented:
Updated so it respects `needsDrop` now. 

Added logic to make it so that produce gets dropped once per day.

Pawns are now ejected at 75% starvation, and a letter is sent. 

Still need to look into the orders/jobs bug, but I think other than potentially adding power usage, we're good! :)

feldoh commented:
Woo, on power usage I don't think that's necessary. Extra ticks and it's a farm, it should work for medieval (meme texture aside 😄). It's basically just a barn imo food can be the only cost

keyz182 commented:
Dang, it's crashing

keyz182 commented:
Ignore, that was me being silly :) All ready now!
